File: 0057-insert-interval/0057-insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        intervals.append(newInterval)
        intervals.sort()
        
        i = 0
        
        while i < len(intervals):
            if i == 0:
                i += 1
                continue
            if intervals[i][0] <= intervals[i - 1][1]:
                if intervals[i][1] > intervals[i - 1][1]:
                    intervals[i - 1][1] = intervals[i][1]
                intervals.pop(i)
                continue
            
            
            
            i += 1
        
        return intervals

## [1,5] [2,5] [6,9]

##[[1,2],[3,5], [4,8],[6,7],[8,10],[12,16]]
## [4,8]

##[[1,2],[3,10],[12,16]]
## [4,8]

File: 0057-insert-interval/0057-insert-interval.rs
use std::cmp::Ordering;

impl Solution {
    pub fn insert(intervals: Vec<Vec<i32>>, new_interval: Vec<i32>) -> Vec<Vec<i32>> {
        let (mut s, mut e) = (new_interval[0], new_interval[1]);

        let pred = |x| {
            move |i: &Vec<i32>| {
                if i[0] > x {
                    Ordering::Greater
                } else if i[1] < x {
                    Ordering::Less
                } else {
                    Ordering::Equal
                }
            }
        };

        let li = intervals.binary_search_by(pred(s)).unwrap_or_else(|x| x);
        let ri = intervals.binary_search_by(pred(e)).unwrap_or_else(|x| x);

        let l = &intervals[..(li + usize::from(li < intervals.len() && intervals[li][1] < s))];
        let r = &intervals[(ri + usize::from(ri >= intervals.len() || intervals[ri][0] <= e))
            .min(intervals.len())..];

        if l.len() + r.len() != intervals.len() {
            s = s.min(intervals[l.len()][0]);
            e = e.max(intervals[intervals.len() - r.len() - 1][1]);
        }

        vec![l, &vec![vec![s, e]], r].concat()
    }
}

// [[1,3],[6,9]] // [2,5]
//
